### DISCUSSION TITLE
Thoughts on Resource Schedueling

### DISCUSSION THREAD

**Discussion by MrPaulAR** [netbox-community/netbox#5860] [General]
We are in the very early stages of adding our topology to Netbox, in our case this is a rather large (think raised floor data centers spanning spread across multiple states) network test & certification environment.  One of our [hopeful] use cases is to allow some of our automations to check in & out a resource (port, device, vlan, compute, etc) or allow a human tester to reserve a resource for a certain period of time.

We initially thought tags would be a good way to do this but there are some limitations on searching tags which looks like it's working as intended by the developers per some issue discussions.   Ideally I would like to be able to search for tag "foo" across all resources and have single view.  I understand I can work around this programmatically and make multiple API calls into netbox.

Are there any other interesting fields that may play into this?  The Tenants may also be interesting for this.  

I don't think netbox has any resource scheduling built in so I understand I would need to create that externally.

**Comment by cpmills1975** [comment]
Could be a common use case for a lab environment I'd suggest. Tenants is the obvious choice, but then you'd lose the ability to identify who owns a device rather than who currently has it allocated to them. Depending on whether you are running the DC for a single company or a multi-tenant hosting environment would decide whether that was a problem or not. Perhaps the tenant could be the person who has it allocated to them and the tenant group is the company they and that asset belongs to?

A more blue-sky approach would be for someone to develop a device allocation/reservation plugin that would add dedicated reservation fields to the object model - perhaps linking an asset with a NetBox user and perhaps adding reservation dates etc. The possibilities are only limited by ones ability to write (or employ someone to write) a Django app.

**Reply by bluikko** [reply]
> tenant group is the company they and that asset belongs to

I might be mistaken but isn't the only field stored for stuff the "tenant" - "tenant group" is not stored.

**Comment by jeremystretch** [comment]
It might help to detail your use case. What does the envisioned workflow look like step-by-step, and what information do you need to track?

**Comment by candlerb** [comment]
I'd find "list all items with a given tag" useful as well.

In the `/extras/tags/` view, it's quite frustrating to see a *count* of the number of objects that have a given tag, which is not clickable. Netbox clearly *knows* which objects have this tag, but won't tell me :-)

So I then have to go and search separately under Devices, Virtual Machines, Inventory Items, Cables etc.  These days, almost everything can be tagged.

I imagine it would work in a very similar way to the global search view - where you get a separate table for each matching class of object, the first few matches, and a clickable link to search just that given class.

(There is a related request #5121, but that's in the opposite direction: being able to filter the tags by associated object type)

**Reply by jeremystretch** [reply]
> I imagine it would work in a very similar way to the global search view - where you get a separate table for each matching class of object, the first few matches, and a clickable link to search just that given class.

The problem with this is that it's _extremely_ inefficient. Each search query results in a separate `SELECT * FROM ...` for each model. We can instead make a single query to the TaggedItem table to find all items assigned to a tag, though we still end up doing a bunch of related queries to retrieve them. (I _think_ we actually used to include this in the original tagging implementation, but ditched it because of performance issues.)

**Reply by candlerb** [reply]
I think the latter approach you give would be more efficient than the current global search.

`select distinct object_id,content_type_id from extras_taggeditem where tag_id=X` immediately gives all the relevant object types and ids.  This means you can limit your queries to just those tables. Then you `select from foo where id in (list_of_ids)` on each table, which uses the primary key.  If global search is acceptably performant, this will be better.

Another option is if `/extras/tags/` breaks down the count, e.g. instead of showing "14" it shows "5 devices, 7 VMs, 2 Inventory Items", and you can click on them individually.  This would be at very little extra cost, with a `group by content_type_id`.

**Comment by MrPaulAR** [comment]
I do expect my use case is quite unique so I really appreciate the interaction here.

Right now this is just a brainstorming session between myself and my mgmt as we are just now discovering this tool and in the very early stages of implementation so I don't have a plan in place.  Today we don't have a system to reserve ports so there isn't a workflow, the question I'm trying to answer is can we do some of this with Netbox and I think the answer is yes so now I'm trying to find out how much.

I can answer some of the questions that have been proposed.

- All of this gear is owned by one business.  Tenancy could be viable instead of groups.
- Reserve (check-out with end date / check-in) interfaces, devices, and patch panel ports (we have smart optical patch panels).  Maybe reserve L2/L3 information as well.

Regarding the reservations this may be as simple as setting a date within a tag or tenant and then I can query on that and do my own stuff outside of netbox to clean up the test environment (tear down optical cross connects, default resources, etc).

**Reply by candlerb** [reply]
There is an accepted feature request on github (#5401) for custom fields on interfaces, which sounds like it could help you with recording interface reservations - it may appear in 2.11.

For now, perhaps you could use a custom field on device, containing a JSON map of port reservations.

**Reply by xkilian** [reply]
Would custom fields with drop down values of (groups/members) be more approriate for what you are doing? Same for the date as a custom field. Custom field models should be available for most things in 2.11.

**Reply by DanSheps** [reply]
@candlerb What do you mean **may** appear: #5401

**Reply by candlerb** [reply]
I did link to #5401.  However I don't believe the code had appeared on `origin/feature` at the time I wrote that :-)

**Comment by MrPaulAR** [comment]
I do think custom fields will be the way go go.  Thanks for the help!

**Reply by drmsoffall** [reply]
I've been thinking about this again recently.  Just curious, how did you end up implementing this?

**Comment by devanshi0114** [comment]


**Reply by candlerb** [reply]
It's a sad day for github when spammers sign up to it :-(

**Comment by MrPaulAR** [comment]
I'm just using tags that I add and remove as needed.
